components/builder-api-client/src/builder.rs

```diff
@@ -1426,6 +1430,45 @@ impl BuilderAPIClient {
     }
 }
 
+/// Retry API operations within reasonable bounds.
+/// Provide an Option with a custom back-off algorithm from the `retry` crate
+/// or a default Fibonacci algorithm will be used. Unrecoverable HTTP errors
+/// will not be retried.
+#[macro_export]
+macro_rules! retry_builder_api {
+    ($iter_opt:expr, $api_future:expr) => {
+        async {
+            let mut iterator = if let Some(i) = $iter_opt {
+                i
+            } else {
+                retry::delay::Fibonacci::from(API_RETRY_WAIT).take(API_RETRIES)
+            };
+            retry::retry_future!(iterator, async {
+                match $api_future.await.into() {
+                    Ok(_) => retry::OperationResult::Ok(()),
+                    Err(e @ api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
+                        println!("Unsupported package platform or architecture. Skipping!");
+                        retry::OperationResult::Ok(())
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::NOT_FOUND, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::UNAUTHORIZED, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::FORBIDDEN, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::UNPROCESSABLE_ENTITY, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e) => retry::OperationResult::Retry(e),
+                }
+            }).await
```

Comment:
At the end we should probably pull out the last error instead of leaving it wrapped in a `retry::OperationResult`. I think you should be able to just use `.error` here.

Reply:
Brilliant! I added in https://github.com/habitat-sh/habitat/pull/7874/commits/2a457e8b6f09ace44d7248a5f5146dcfbd2bb44e